src/jit/codegenarm64.cpp

```diff
@@ -862,18 +874,90 @@ void                CodeGen::genFuncletProlog(BasicBlock* block)
+
+        // First instruction
+        // Load the CallerSP of the main function (stored in the PSP of the dynamically containing funclet or function)
+        if (emitter::emitIns_valid_imm_for_ldst_offset(genFuncletInfo.fiCallerSP_to_PSP_slot_delta, EA_8BYTE))
+        {
+            getEmitter()->emitIns_R_R_I(ins_Load(TYP_I_IMPL), EA_PTRSIZE, REG_R1, 
+                                        REG_R1, genFuncletInfo.fiCallerSP_to_PSP_slot_delta);
+            regTracker.rsTrackRegTrash(REG_R1);
+        }
+        else
+        {
+            // Setup X2 with the large constant offset
+            instGen_Set_Reg_To_Imm(EA_PTRSIZE, REG_R2, genFuncletInfo.fiCallerSP_to_PSP_slot_delta);
+            regTracker.rsTrackRegTrash(REG_R2);
+            
+            // Load the CallerSP
+            getEmitter()->emitIns_R_R_R(ins_Load(TYP_I_IMPL), EA_PTRSIZE, REG_R1, REG_R1, REG_R2);
+            regTracker.rsTrackRegTrash(REG_R1);
+        }
+
+        // Second instruction
+        // Store the PSP value (aka CallerSP)
+        if (emitter::emitIns_valid_imm_for_ldst_offset(genFuncletInfo.fiSP_to_PSP_slot_delta, EA_8BYTE))
+        {            
+            getEmitter()->emitIns_R_R_I(ins_Store(TYP_I_IMPL), EA_PTRSIZE, REG_R1, 
+                                        REG_SPBASE, genFuncletInfo.fiSP_to_PSP_slot_delta);
+        }
+        else
+        {
+            // Setup X2 with the large constant offset
+            instGen_Set_Reg_To_Imm(EA_PTRSIZE, REG_R2, genFuncletInfo.fiSP_to_PSP_slot_delta);
+            regTracker.rsTrackRegTrash(REG_R2);
+            
+            // Store the PSP value
+            getEmitter()->emitIns_R_R_R(ins_Store(TYP_I_IMPL), EA_PTRSIZE, REG_R1, REG_SPBASE, REG_R2);
+        }
 
-        getEmitter()->emitIns_R_R_I(ins_Load(TYP_I_IMPL), EA_PTRSIZE, REG_R1, REG_R1, genFuncletInfo.fiCallerSP_to_PSP_slot_delta);
-        regTracker.rsTrackRegTrash(REG_R1);
-        getEmitter()->emitIns_R_R_I(ins_Store(TYP_I_IMPL), EA_PTRSIZE, REG_R1, REG_SPBASE, genFuncletInfo.fiSP_to_PSP_slot_delta);
-        getEmitter()->emitIns_R_R_I(INS_add, EA_PTRSIZE, REG_FPBASE, REG_R1, genFuncletInfo.fiFunction_CallerSP_to_FP_delta);
+        // Third instruction
+        // re-establish the frame pointer
+        if (emitter::emitIns_valid_imm_for_add(genFuncletInfo.fiFunction_CallerSP_to_FP_delta, EA_8BYTE))
+        {
+            getEmitter()->emitIns_R_R_I(INS_add, EA_PTRSIZE, REG_FPBASE, 
+                                        REG_R1, genFuncletInfo.fiFunction_CallerSP_to_FP_delta);
+        }
+        else
+        {
+            // Setup X2 with the large constant offset
+            instGen_Set_Reg_To_Imm(EA_PTRSIZE, REG_R2, genFuncletInfo.fiFunction_CallerSP_to_FP_delta);
+            regTracker.rsTrackRegTrash(REG_R2);
+            
+            // re-establish the frame pointer
+            getEmitter()->emitIns_R_R_R(INS_add, EA_PTRSIZE, REG_FPBASE, REG_R1, REG_R2);
+        }
```

Comment:
We have a lot of this pattern:

if (constant fits) {
    generate one instruction using constant
} else {
   load constant to temp register
   generate instruction using temp register
}

I wonder if we could abstract this with a function like:

genInstrWithConstant(ins, type, reg1, reg2, const, regTemp, constant_size_checking_function)